arch/risc-v/src/mpfs/mpfs_usb.c

```diff
@@ -0,0 +1,3770 @@
+
+enum mpfs_epstate_e
+{
+  USB_EPSTATE_DISABLED = 0, /* Endpoint is disabled */
+  USB_EPSTATE_STALLED,      /* Endpoint is stalled */
+  USB_EPSTATE_IDLE,         /* Endpoint is idle */
+  USB_EPSTATE_SENDING,      /* Endpoint is sending data */
+  USB_EPSTATE_RXSTOPPED,    /* EP is stopped waiting for a read request */
+  USB_EPSTATE_EP0DATAOUT,   /* Endpoint 0 is receiving SETUP OUT data */
+  USB_EPSTATE_EP0STATUSIN,  /* Endpoint 0 is sending SETUP status */
+  USB_EPSTATE_EP0ADDRESS    /* Address change is pending completion */
+};
+
+/* Device states */
+
+enum mpfs_devstate_e
+{
+  USB_DEVSTATE_SUSPENDED = 0, /* The device is currently suspended */
+  USB_DEVSTATE_POWERED,       /* Host is powered through the USB cable */
+  USB_DEVSTATE_DEFAULT,       /* Device has been reset */
+  USB_DEVSTATE_ADDRESSED,     /* The device has an address on the bus */
+  USB_DEVSTATE_CONFIGURED     /* A valid configuration has been selected. */
+};
+
+/* The result of EP0 SETUP */
+
+enum mpfs_ep0setup_e
+{
+  USB_EP0SETUP_SUCCESS = 0,   /* The SETUP was handled without incident */
+  USB_EP0SETUP_DISPATCHED,    /* The SETUP was forwarded */
+  USB_EP0SETUP_ADDRESS,       /* A new device address is pending */
+  USB_EP0SETUP_STALL          /* An error occurred */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int    mpfs_ep_configure(struct usbdev_ep_s *ep,
+                const struct usb_epdesc_s *desc, bool last);
+static int    mpfs_ep_disable(struct usbdev_ep_s *ep);
+static struct usbdev_req_s *
+              mpfs_ep_allocreq(struct usbdev_ep_s *ep);
+#ifdef CONFIG_USBDEV_DMA
+static void  *mpfs_ep_allocbuffer(struct usbdev_ep_s *ep, uint16_t nbytes);
+static void   mpfs_ep_freebuffer(struct usbdev_ep_s *ep, void *buf);
+#endif
+static void   mpfs_ep_freereq(struct usbdev_ep_s *ep,
+                struct usbdev_req_s *);
+static int    mpfs_ep_submit(struct usbdev_ep_s *ep,
+                struct usbdev_req_s *req);
+static int    mpfs_ep_cancel(struct usbdev_ep_s *ep,
+                struct usbdev_req_s *req);
+static int    mpfs_ep_stallresume(struct usbdev_ep_s *ep, bool resume);
+
+/* USB device controller operations */
+
+static struct usbdev_ep_s *
+              mpfs_allocep(struct usbdev_s *dev, uint8_t epno, bool in,
+                uint8_t eptype);
+static void   mpfs_freeep(struct usbdev_s *dev, struct usbdev_ep_s *ep);
+static int    mpfs_getframe(struct usbdev_s *dev);
+static int    mpfs_wakeup(struct usbdev_s *dev);
+static int    mpfs_selfpowered(struct usbdev_s *dev, bool selfpowered);
+static int    mpfs_pullup(struct usbdev_s *dev,  bool enable);
+
+/* Interrupt level processing */
+
+static void   mpfs_ep0_ctrlread(struct mpfs_usbdev_s *priv);
+static void   mpfs_ep0_wrstatus(struct mpfs_usbdev_s *priv,
+                const uint8_t *buffer, size_t buflen);
```

Comment:
Can we align new lines with `(`?

Reply:
Now I read the comments, I'll address the suggestions on a separate patch (as the request appears merged). Thanks for the suggestions.